Engine/src/Engine/ResourceManagement/AdvancedShaderDistributor.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+#ifndef ADVANCEDSHADERDISTRIBUTOR_H
+#define ADVANCEDSHADERDISTRIBUTOR_H
+
+#include <string>
+#include <map>
+#include <vector>
+#include <Engine/ResourceManagement/ShaderFillType.h>
+#include <Engine/ResourceManagement/ShaderGenerator.h>
+
+namespace Engine
+{
+	// Based on the advanced shader source code, this generates shaders for each
+	// shader fill type (check ShaderFillType.h). This means that each advanced
+	// shader can be applied to different sets of vertex data such as gradients,
+	// flat colours, and sprites. Above all, it makes it easier for the user
+	// developing the advanced shader since the advanced shader code and the
+	// default shader code are separate.
+	class AdvancedShaderDistributor
+	{
+	public:
+		AdvancedShaderDistributor(std::string advancedSource,
+			std::map<ShaderFillType::FillType, ShaderGenerator>& shaders);
+		//returns programId to run the shader. Note that advanced shaders are combined with each
+		//default shader, so the corresponding default shader enum must be assigned.
+		GLuint getAdvancedProgramId(ShaderFillType::FillType fillType = ShaderFillType::FillType::SN_TEXTURE_FILL);
+
+	private:
+		const static inline int SHADER_VERSION = 330;
+		GLuint advProgramIds[ShaderFillType::types];
+		std::string advancedSource;
+		std::string vertexSourceSet[ShaderFillType::types], fragmentSourceSet[ShaderFillType::types];
+		const std::vector<std::string> VARIABLE_TYPES
+		{ "sampler2D", "bool", "int", "uint", "float", "double", "bvec2", "bvec3", "bvec4", "ivec2", "ivec3",
+			"ivec4", "uvec2", "uvec3", "uvec4", "vec2", "vec3", "vec4", "dvec2", "dvec3", "dvec4",
+			"mat2", "mat3", "mat4"};
+		struct SourceExtractor
+		{
+			std::vector<std::string> layout;
+			std::vector<std::string> inVariables;
+			std::vector<std::string> outVariables;
+			std::vector<std::string> uniforms;
+			std::vector<std::string> mainCommands;
+
+			//Prints the variables. Only used for testing purposes
+			void print();
+		};
+		enum SEMessenger
+		{
+			PENDING_NONE,
```

Comment:
Could we add a prefix here to keep with the pseudo-convention some other enums in our codebase have?

Reply:
Done